omega/DecisionTree/DecisionTreeClassifier.py

```diff
@@ -0,0 +1,230 @@
+    >>> dt.fit(x, y)
+    >>> dt.predict(torch.Tensor([10.12493903, 3.234550982]))
+    1.0
+    """
+
+    def __init__(self,
+                 max_depth=100,
+                 min_size=2,
+                 split_metric='gini_index'):
+        """Initializes the parameters of a decision tree and create root node
+        the decision as an empty dictionary.
+
+        Parameters
+        ----------
+        max_depth : int, optional, default value 100
+            Represents maximum allowed depth of the binary tree.
+        min_size : int, optional, default value 2
+            Represents minimum number of data points associated with each node.
+        split_metric : str, optional, default value gini_index
+            possible values - gini_index, entropy_loss(under development)
+            Represents the metrics to compare each split point.
+        """
+        self.__use_cuda = torch.cuda.is_available()
+        self.__max_depth = max_depth
+        self.__min_size = min_size
+        self.__split_metric = split_metric
+        self.__root = {}
+
+    def __gini_index(self, left, right, classes):
+        """Calculates the value of gini index for each split.
+
+        Parameters
+        ----------
+        left :  a fraction of the original data
+        right : a fraction of the original data
+        classes : list of integers
+            list of all possible values to target labels.
+        """
+        num_instances = sum([len(left[1]), len(right[1])])
+        gini = 0.0
+        for group in [left, right]:
+            y = group[1]
+            size = len(y) * 1.0
+            if size == 0:
+                continue
+            score = 0.0
+            for cl in classes:
+                proportion = sum(y == cl) / size
+                score += proportion ** 2
+            gini += (1.0 - score) * (size / num_instances)
+        return gini
+
+    def __test_split(self, index, value, x, y):
+        """Splits the dataset into two parts on the basis of a value.
+
+        Parameters
+        ----------
+        index :  index of the attribute on which to split
+        value : attribute value on which to split
+        x : training data
+        y : training labels
+        """
+        mask = torch.nonzero(torch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            left = (torch.index_select(x, 0, mask), torch.index_select(y, 0, mask))
+        else:
+            left = (torch.DoubleTensor(), torch.DoubleTensor())
+
+        mask = torch.nonzero(1 - torch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            right = (torch.index_select(x, 0, mask), torch.index_select(y, 0, mask))
+        else:
+            right = (torch.DoubleTensor(), torch.DoubleTensor())
+        return left, right
+
+    def fit(self, x, y):
+        """Trains the classifier on the training data and forms the decision tree
```

Comment:
Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html